MTL-2348 CSM 1.4.4 Patch (#4647)
Cray-HPE/docs-csm: operations/configuration_management/Management_Node_Image_Customization.md

    1. Get the existing `metal.server` setting for the component name (xname) of the node of interest.

       ```bash
       XNAME=<node-xname>
       METAL_SERVER=$(cray bss bootparameters list --hosts "${XNAME}" --format json | jq '.[] |."params"' \
           | awk -F 'metal.server=' '{print $2}' \
           | awk -F ' ' '{print $1}')
       echo "${METAL_SERVER}"
       ```

    1. Create updated boot parameters that point to the new artifacts.

       1. Set the path to the artifacts in S3.

          ***NOTE*** This uses the `NEW_IMS_IMAGE_ID` variable set in an earlier step.

          ```bash
          S3_ARTIFACT_PATH="boot-images/${NEW_IMS_IMAGE_ID}"
          echo "${S3_ARTIFACT_PATH}"
          ```

       1. Set the new `metal.server` value.

          ```bash
          NEW_METAL_SERVER="s3://${S3_ARTIFACT_PATH}/rootfs"
          echo "${NEW_METAL_SERVER}"
          ```

       1. Determine the modified boot parameters for the node.

          ```bash
          PARAMS=$(cray bss bootparameters list --hosts "${XNAME}" --format json | jq '.[] |."params"' | \
              sed "/metal.server/ s|${METAL_SERVER}|${NEW_METAL_SERVER}|" | \

>               sed "s/metal.no-wipe=1/metal.no-wipe=0/" | \

Reviewer: Remove this line. No wipe is automatically set via workflows. Here it is set in worker node upgrade: https://github.com/Cray-HPE/docs-csm/blob/release/1.4/workflows/ncn/worker/worker.wipe-and-reboot.yaml#L59. Here is where it is set in master node upgrade: https://github.com/Cray-HPE/docs-csm/blob/release/1.4/upgrade/scripts/upgrade/ncn-upgrade-master-nodes.sh#L163.

Reviewer: This was recently removed in CSM 1.5 and CSM 1.6 
docs-csm 1.6 PR: https://github.com/Cray-HPE/docs-csm/pull/4626
docs-csm 1.5 PR: https://github.com/Cray-HPE/docs-csm/pull/4627

Author: is this section strictly for masters and workers? I think this is the manual version of the workflow, and caters to all NCNs.

Reviewer: That section is for all NCNs. I see that the 1.4.4 patch doesn't use that section. I do think it should be removed since metal.no-wipe in each node's rebuild/upgrade workflow. However, it probably won't cause a problem to leave. I do think you need to remove the metal.no-wipe in the assign-ncn-images script because we don't want to set it on all nodes, especially storage nodes.

Author: I think it might need to be left in, because if someone is following this guide then they aren't using the workflow.

As for the script, I think you found a good bug. That script shouldn't be setting that, it is because I copied it from the docs. That script should only be updating the image, initrd, and kernel URLs.

That means I'll need a new step where users set `metal.no-wipe` for this upgrade from 1.4.x to 1.4.4.

    * Get a comma-separated list of all worker NCN xnames:

        ```bash
        if IFS=$'\n' read -rd '' -a XNAMES; then
          :

>         fi <<< "$(cray hsm state components list --role Management --subrole Worker --type Node --format json | jq -r '.Components | map(.ID) | join("\n")')"

Reviewer 2: As the preceding text indicates, the original command here produced a comma-separated list of worker xnames. The new command produces no output. I believe your new command is intended to be used with the script in the subsequent steps. If so, the text above should probably be adjusted accordingly.

Alternatively, you could just add an additional line to your new command:
```bash
echo "Found the following worker NCNs: ${XNAMES}"
```

That way it would also let them know if the command found no workers, or a weird amount of workers.

Author: Yes, it will need rewording.
I changed this because frankly I couldn't believe I was going to have to run these steps umpteen times (since I am not using IUF for this).

        ```

    * Get a comma-separated list of all master NCN xnames:

        ```bash
        if IFS=$'\n' read -rd '' -a XNAMES; then
          :

>         fi <<< "$(cray hsm state components list --role Management --subrole Master --type Node --format json | jq -r '.Components | map(.ID) | join("\n")')"

Reviewer 2: Same issue as with the worker NCNs

        ```

    * Get a comma-separated list of all storage NCN xnames:

        ```bash
        if IFS=$'\n' read -rd '' -a XNAMES; then
          :

>         fi <<< "$(cray hsm state components list --role Management --subrole Storage --type Node --format json | jq -r '.Components | map(.ID) | join("\n")')"

Reviewer 2: Same issue as with the worker NCNs

1. (`ncn-mw#`) Update boot parameters for a management node.

>     1. When working with a group of `XNAMES` (use the previous step to collect the group of `XNAMES`).

Reviewer 2: This shouldn't be a numbered list since you do this option OR the next option, not do them in sequence, if I am understanding the new flow correctly.

```suggestion
    * When working with a group of xnames (use the previous step to collect the xnames in the `$XNAME` variable).
```

        ```bash
        for XNAME in "${XNAMES[@]}"; do
          echo "--- $XNAME ---"
          METAL_SERVER=$(cray bss bootparameters list --hosts "${XNAME}" --format json | jq '.[] |."params"' \
            | awk -F 'metal.server=' '{print $2}' \
            | awk -F ' ' '{print $1}')
          echo "current metal.server: ${METAL_SERVER}"
          S3_ARTIFACT_PATH="boot-images/${NEW_IMS_IMAGE_ID}"
          NEW_METAL_SERVER="s3://${S3_ARTIFACT_PATH}/rootfs"
          echo "    new metal.server: ${NEW_METAL_SERVER}"
          PARAMS=$(cray bss bootparameters list --hosts "${XNAME}" --format json | jq '.[] |."params"' | \
            sed "/metal.server/ s|${METAL_SERVER}|${NEW_METAL_SERVER}|" | \
            sed "s/metal.no-wipe=1/metal.no-wipe=0/" | \
            tr -d \")
          cray bss bootparameters update --hosts "${XNAME}" \
            --kernel "s3://${S3_ARTIFACT_PATH}/kernel" \
            --initrd "s3://${S3_ARTIFACT_PATH}/initrd" \
            --params "${PARAMS}"
        done

>         ```

Reviewer 2: I don't see any return codes being checked in this code snippet. If it hits a problem on the first xname but the rest work fine, the error messages will be at the top of the output. An admin who only looks at the tail end of the output would possibly overlook the earlier failures. Ideally it should check for failing commands and break out of the loop immediately when one is hit, with an appropriate error message.

This may make more sense to be moved into a separate shell script file, for this reason.

Reviewer 2: This comment wasn't addressed in your earlier pass. Not sure if it was overlooked.

Author: I did this rather impatiently, because I couldn't believe the steps were having me run this once-per NCN before. I can shell-script this.

>     1. When working on one node at a time.

Reviewer 2: ```suggestion
    * When working on one node at a time.
```

                S3_ARTIFACT_PATH="boot-images/${NEW_IMS_IMAGE_ID}"
                echo "${S3_ARTIFACT_PATH}"
                ```

            1. Set the new `metal.server` value.

                ```bash
                NEW_METAL_SERVER="s3://${S3_ARTIFACT_PATH}/rootfs"
                echo "${NEW_METAL_SERVER}"
                ```

            1. Determine the modified boot parameters for the node.

                ```bash
                PARAMS=$(cray bss bootparameters list --hosts "${XNAME}" --format json | jq '.[] |."params"' | \
                    sed "/metal.server/ s|${METAL_SERVER}|${NEW_METAL_SERVER}|" | \
                    sed "s/metal.no-wipe=1/metal.no-wipe=0/" | \
                    tr -d \")
                echo "${PARAMS}"
                ```

                In the output of the `echo` command, verify that the value of `metal.server` is
                correctly set to the value of `${NEW_METAL_SERVER}`.

        1. Update BSS with the new boot parameters.

            ```bash
            cray bss bootparameters update --hosts "${XNAME}" \
                --kernel "s3://${S3_ARTIFACT_PATH}/kernel" \
                --initrd "s3://${S3_ARTIFACT_PATH}/initrd" \

>                 --params "${PARAMS}"

Reviewer 2: ```suggestion
                --params "${PARAMS}" && echo SUCCESS || echo ERROR
```

Reviewer 2: This comment wasn't addressed in your earlier pass. Not sure if it was overlooked.

1. (`ncn-mw#`) Update boot parameters for a management node.

>     * When working with a group of `XNAMES` (use the previous step to collect the group of `XNAMES`).

Reviewer 2: ```suggestion
    * When working with a group of xnames (use the previous step to collect the xnames in the `$XNAME` variable).
```

        ```bash
        for XNAME in "${XNAMES[@]}"; do
          echo "--- $XNAME ---"
          METAL_SERVER=$(cray bss bootparameters list --hosts "${XNAME}" --format json | jq '.[] |."params"' \
            | awk -F 'metal.server=' '{print $2}' \
            | awk -F ' ' '{print $1}')
          echo "current metal.server: ${METAL_SERVER}"
          S3_ARTIFACT_PATH="boot-images/${NEW_IMS_IMAGE_ID}"
          NEW_METAL_SERVER="s3://${S3_ARTIFACT_PATH}/rootfs"
          echo "    new metal.server: ${NEW_METAL_SERVER}"
          PARAMS=$(cray bss bootparameters list --hosts "${XNAME}" --format json | jq '.[] |."params"' | \
            sed "/metal.server/ s|${METAL_SERVER}|${NEW_METAL_SERVER}|" | \
            sed "s/metal.no-wipe=1/metal.no-wipe=0/" | \
            tr -d \")
          cray bss bootparameters update --hosts "${XNAME}" \
            --kernel "s3://${S3_ARTIFACT_PATH}/kernel" \
            --initrd "s3://${S3_ARTIFACT_PATH}/initrd" \
            --params "${PARAMS}"
        done
        ```

    * When working on one node at a time.

        1. Get the existing `metal.server` setting for the component name (xname) of the node of interest.

            ```bash

>             XNAME=<node-xname>

Reviewer: Is there a reason that the once the XNAME is set here, you did not use the single block of code that was used in a loop above to execute all these steps together?

Author: I had left the original steps alone, which were for running against one node at a time.
They could be enhanced, I think it would be better/simpler if they were.

Reviewer: Makes sense. It works as is. We can make this change in a new PR that will go into all the branches.

Author: Opting for that, marking this as resolved.
